src/audio/kpb.c

```diff
@@ -119,11 +123,212 @@ static void kpb_free(struct comp_dev *dev)
+				"KPB bufefrs");
+		return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr +
+	(MAX_BUFFER_SIZE - LPSRAM_SIZE);
+
+	cd->his_buf_lp.end_addr = cd->his_buf_lp.sta_addr +
+	LPSRAM_SIZE;
+
+#elif KPB_NO_OF_HISTORY_BUFFERS == 1
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE);
+
+	if (cd->his_buf_hp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+	return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr + MAX_BUFFER_SIZE;
+
+#else
+#error "Wrong number of key phrase buffers configured"
+#endif
+
+	/* TODO: zeroes both buffers */
+
+	/* Initialize clients data */
+	for (i = 0; i < MAX_NO_OF_CLIENTS; i++)
+		cd->clients[i].state = KPB_CLIENT_UNREGISTERED;
+
+	return ret;
+}
+
+/**
+ * kpb_copy() - copy real time input stream into sink buffer,
+ * and in the same time buffers that input for
+ * later use by some of clients.
+ *
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_copy(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+
+	trace_kpb("kpb_copy()");
+
+	/* Get source and sink buffers */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 source_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       sink_list);
+	/* Process source data */
+	/* Check if source have data and valid pointer to read from */
+	if (source->avail > 0 && source && sink && source->r_ptr) {
+		/* Real time copying */
+		if (sink->free >= source->avail && sink->w_ptr) {
+			/* Sink and source are both ready and have space */
+			memcpy(sink->w_ptr, source->r_ptr,
+			       source->avail);
+
+			/* update source & snik data*/
+			comp_update_buffer_produce(sink, source->avail);
+			comp_update_buffer_consume(source, source->avail);
+		} else {
+			;/* What should we do if sink isn't ready(state/size)?
+			  * Simple skip is OK?
+			  */
+		}
+		/* Buffer source data internally for future use by clients */
+		if (source->avail <= MAX_BUFFER_SIZE)
+			kpb_buffer_data(kpb, source);
+	} else
+		ret = -EINVAL;
+
+	return ret;
+}
+
+static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source)
+{
+}
+
+/**
+ * kpb_cmd() - kpb event handler.
+ * @arg1:  kpb component.
+ * @arg2:  command to be executed
+ * @arg3:  pointer to data related with request
+ * @arg4:  maxiumum size of data
+ *
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	int ret = 0;
+	struct sof_ipc_ctrl_data *cdata = data;
+	struct kpb_client *cli = (struct kpb_client *)cdata->data->data;
+	uint8_t client_id = cli->id;
+
+	trace_kpb("kpb_cmd()");
+
+	switch (cmd) {
+	case KPB_EVENT_REGISTER_CLIENT:
```

Comment:
@mrajwa I feel that this step might unnecessary. I am thinking of something like a MUX which is connected to 5 sink components. This information can be obtained from topology. And can be retriedved by looking at the sink buffer list for the component.

Reply:
I am not sure, KPB HLD documents specifies this client registration procedure. Keep in mind that every sink other than the real time one, needs to register in KPB prior to its draining request. This is obvious since at draining time KPB must know how much data shall be drained. I can see the use of MUX which may specify sink as one ofthe buffer on bsink_list and KPB cold fetch history depth as sink->free - is that your proposition about? 
However I think this client oriented approach is much cleaner and allows us to easy extend KPB functionality later on if we decide to add more pre checks before draining.

Reply:
@mrajwa is HLD = high level design? Can you point me to these documents?

Reply:
@dbaluta, I don't know its rather for internal use only. I will ask the owner of it whether it can be shared.
Anyway, I need to agree with driver side as for how are we going to communicate for events like "draining". I can imagine handle this scenario without clients/sinks registrations etc (although I do like this idea) but we need to make some decisions:
1. What .ops will be used to start draining for specific sink/client.
2. How draining parameters like "history_depth" will be passed into KPB and how they can be modify, if at all.
3. How to communicate failure/success of this procedure, if at all.

@plbossart, @xiulipan, @keyonjie could you please answer above questions?